### ISSUE TITLE
Chapter 2: Issue and discussion regarding using dataclasses and SQLAlchemy

### ISSUE THREAD

**Issue by sfermigier** [cosmicpython/book#154]
Hi,

I've tried to implement the approach you present in Chapter 2 in a recent project, instead of the more traditional ORM Declarative way, and hit a bug that I discussed on the SQLAlchemy bug tracker:

https://github.com/sqlalchemy/sqlalchemy/issues/5027

In reply, Mike Bayer said that he didn't intend for the SQLAlchemy classical mapper to work with dataclasses (in other words, that it's unsupported) and an interesting discussion ensued.

**Comment by hjwp** [comment]
yes.  I did have a discussion with MIke when we first started on the book, because I was hacking around with `mapper()` and finding it a bit hard and less well documented, and he told me he wasn't so hot on the concept any more.  I can see where he's coming from, although i don't quite agree with everything he says in that thread in particular -- you can see he's still in a mentality of assuming the database and the ORM come first, rather than the domain model coming first.  His example of an `Order` model that has a `get_order_total_message()` method that depends on a `dollar_amount` that he thinks the class would have to "magically assume" is already there.  In our world of course, the class would not magically assume anything, it would have an init that required dollar_amount, and it would have a whole suite of unit tests that are totally independent of the ORM.

fwiw, in my last project we started out without an ORM, just writing raw SQL in our repositories, just for fun. we assumed it eventually would grow too tedious to maintain hand-rolled sql and we'd want to switch to an ORM, but when we actually tried it, we found it actually created more problems than it solved -- ever seen those "this instance is no longer attached to a session" errors?

I've spoken to other people who say they haven't used an ORM in a decade.  maybe we should spend a little time in the book mentioning that it's a viable option.

**Comment by hjwp** [comment]
How are you liking the book in general Stefane?  How far through are you?  I remember you commented on my blog post about "architecture stuff" from way back when we were starting... any reactions to what you've seen so far?  We go to print in january, so very keen to get any suggestions in before then!
